src/logging/wal_logger.cpp

```diff
@@ -0,0 +1,101 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// wal_logger.cpp
+//
+// Identification: src/logging/wal_logger.cpp
+//
+// Copyright (c) 2015-18, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include <fstream>
+#include "logging/wal_logger.h"
+#include "logging/log_buffer.h"
+#include "catalog/manager.h"
+#include "common/container_tuple.h"
+#include "storage/tile_group.h"
+#include "../include/logging/wal_logger.h"
+
+
+namespace peloton{
+namespace logging{
+
+bool WalLogger::IsFlushNeeded(bool pending_buffers){
+
+  /* if the disk buffer is full then we flush to the disk
+   * irrespective of whether there are pending callbacks
+   */
+  if(disk_buffer_->HasThresholdExceeded())
+    return true;
+
+  /* if the disk buffer is not full and there are pending buffers
+   * in the queue then we don't flush to the disk now. We try to
+   * consolidate the pending buffers to achieve group commit.
+   */
+  if(pending_buffers)
+    return false;
+
+  /* if there are no pending buffers (at the time of check), but
+   * there are pending callbacks, we go ahead and flush the
+   * buffers.
+   */
+  if(!callbacks_.empty())
+    return true;
+
+  return false;
+}
+
+
+void WalLogger::FlushToDisk(){
+
+  if(disk_buffer_->GetSize()==0)
+    return;
+
+  std::ofstream *stream = LogManager::GetInstance().GetFileStream();
+  stream->write(disk_buffer_->GetData(), disk_buffer_->GetSize());
```

Comment:
Multiple logger threads connecting with the same output stream may result in race condition and corrupt the log files. Maybe consider locking the file while writing and flush.

Reply:
The plan as of now and in the foreseeable future is to just use one logger thread. The locking changes can be made as a part of another PR. For this PR, we are only using one logger thread.